### ISSUE TITLE
`__iter__` implementation not detected in superclass of Pyomo OrderedScalarSet

### ISSUE THREAD

**Issue by alexchandel** [microsoft/pylance-release#4667]
## Environment data

-   Language Server version: 2023.7.40
-   OS and version: macOS 13.4.1
-   Python version (& distribution if applicable, e.g. Anaconda): 3.11.4

## Code Snippet

```python
#!/usr/bin/env python3
"""Test."""
from pyomo.environ import ConcreteModel, Set
from pyomo.core.base.set import OrderedScalarSet


class MyModel(ConcreteModel):
    """Model."""

    c: OrderedScalarSet

    def __init__(self, comps):
        """Init."""
        super().__init__()
        self.c = Set(initialize=tuple(comps), doc='component')  # Set() returns an OrderedScalarSet, which Pylance also fails to detect


model = MyModel(['MEOH', 'IB4', 'NB4', 'MTBE'])
print([model.c, type(model.c)])  # OrderedScalarSet
x = {k: 1 for k in model.c}  # Pylance falsely reports OrderedScalarSet has no __iter__
```

## Expected behavior

No error on iteration. `OrderedScalarSet` implements `__iter__` via its ancestor  `_SetData` (via `_InsertionOrderSetData` → `_OrderedSetData` → `_FiniteSetData` → `_SetData`).

## Actual behavior

```
"OrderedScalarSet" is not iterable
  "__iter__" method does not return an object
```

**Comment by erictraut** [comment]
Pyomo is an untyped library — that is, it doesn't contain any type annotations. I presume that you have not installed any type stubs for it. That means pylance will attempt to infer types from the source if possible. Type inference works pretty well in most cases, but there are situations where incorrect types will be inferred. If this library is important to you, consider logging a request with the maintainers to add inlined type annotations. Many libraries are adding type information to assist with type checking and language server features like completion suggestions. 

The problem here is that the `__iter__` method is supplied by the `_FiniteSetMixin` class, which is defined in the `pyomo/core/base/set.py` source file. The `__iter__` method doesn't have a return type annotation, so its return type is inferred based on the return statement `return self._iter_impl()`.  The `_iter_impl` method likewise has no return type annotation. The implementation of this method is:

```python
    def _iter_impl(self):
        raise DeveloperError("Derived finite set class (%s) failed to "
                             "implement _iter_impl" % (type(self).__name__,))
```

The return type is therefore inferred as `NoReturn` which indicates that the method raises an exception on all code paths. There are three ways that the library maintainer can fix this:
1. Provide an explicit type annotation for `__iter__` or `_iter_impl`.
2. Mark the `_FiniteSetMixin` class as an abstract base class (i.e. add `abc.ABC` as a base class) and mark the `_iter_impl` method as `@abstractmethod`.
3. Change the raised exception type from `DeveloperError` to `NotImplemented`. Pylance special-cases this exception because it's commonly used to indicate that a method must be implemented by a subclass.

All three of these are considered best practices, but any one of them will be sufficient to fix the problem. If the library maintainer has questions, we're happy to help.

If you find that you're seeing too many false positive type checking errors caused by the lack of type information in this library, you have a couple of options:
1. Disable all type checking diagnostics by setting `python.analysis.typeCheckingMode` to "off".
2. Add a `# type: ignore` comment at the end of the line to suppress a single diagnostic.
3. Set `python.analysis.useLibraryCodeForTypes` to false. This will tell pylance not to attempt to infer any types from untyped library code. This will eliminate false positive errors, but you will lose all completion suggestions and other language server features for classes and methods that come from the library.

**Comment by erictraut** [comment]
Digging a bit deeper, I think your assumption about the `Set` constructor returning an instance of `OrderedScalarSet` is incorrect. If you change the declared type of `c` from `OrderedScalarSet` to `Set`, all of your type errors disappear. That's because `Set` has an `__iter__` method (supplied by base class `IndexedComponent`) that returns a valid iterable object.

**Comment by alexchandel** [comment]
You're correct that I'm relying on Pylance's type inference at the moment.  There are just a couple classes of false positives, like this one (of which there are unfortunately many instances of), so I'd rather fix or work around the issue than disable type inference, as it's very useful outside those false positives.

> Digging a bit deeper, I think your assumption about the `Set` constructor returning an instance of `OrderedScalarSet` is incorrect. If you change the declared type of `c` from `OrderedScalarSet` to `Set`, all of your type errors disappear. That's because `Set` has an `__iter__` method (supplied by base class `IndexedComponent`) that returns a valid iterable object.

Run the code yourself, you will see it prints that the type is `OrderedScalarSet`.  I have to declare it as `OrderedScalarSet` because the larger codebase uses methods that `Set` doesn't have, but `OrderedScalarSet` does, like `.at()`.

Re: `_FiniteSetMixin` having a NoReturn `_iter_impl`, while that is true, the child `_FiniteSetData` does have an implementation of `_iter_impl`, and that's the implementation that `OrderedScalarSet` ends up using.  Shouldn't Pylance detect this?

**Comment by alexchandel** [comment]
I note that `DeveloperError` is subclass of `NotImplementedError`.  Isn't that enough to tell Pylance to look for another implementation among the descendants?

**Comment by alexchandel** [comment]
[Raising NotImplementedError is suggested as a best practice for abstract methods here](https://stackoverflow.com/a/44575926), so I would assume that applies to exceptions that inherit from `NotImplementedError`.

**Comment by erictraut** [comment]
Pyright (the type checker upon which pyright is based) is working as intended here, so I don't consider this a bug. See [this documentation](https://microsoft.github.io/pyright/#/type-inference?id=noreturn-return-type) for details.

If you would like to file an enhancement request in the [pyright project](https://github.com/microsoft/pyright) to support subclasses of `NotImplementedError`, I think that's a reasonable request. This is the first time I've seen a library subclass `NotImplementedError`, so this isn't common. But this should be relatively easy and cheap to support.

Looking at the library code, I'm not sure how a call to the `Set` constructor is returning a `OrderedScalarSet` object. It must be doing some dynamic hackery if that's the case. Static type analyzers will not be able to infer this without the assistance of type annotations.

**Comment by alexchandel** [comment]
Ok, I opened a request for it: microsoft/pyright#5608

It might be some kind of dynamic hackery, via `return super(Set, cls).__new__(AbstractOrderedScalarSet)`, which gets invoked because `ordered` defaults to `Set.InsertionOrder`.  Pyomo does extremely dynamic things in other places, like [how it declares](https://github.com/Pyomo/pyomo/blob/main/pyomo/core/base/set.py#L4412C12-L4412C12) the types for `from pyomo.environ import Binary, NonNegativeReals`.  It's very helpful, because Pyomo turns Python into a DSL for a large number of optimizers, but unfortunately requires explicit hints.

I think the maintainer would be willing to accept type inferences on things like the Set constructor, but I honestly have no idea how to annotate that without dependent types.  I'm using a `cast` at the moment (omitted from the sample).

**Comment by alexchandel** [comment]
I get a seemingly similar error when iterating over Pyomo's `Var`:

```py
#!/usr/bin/env python3
"""Test."""
from pyomo.environ import Binary, ConcreteModel, Var
from pyomo.core.base.var import IndexedVar


class MyModel(ConcreteModel):
    """Model."""

    y_Ip: IndexedVar

    def __init__(self, ns: tuple[str, str], ps: tuple[str, str]):
        """Init."""
        super().__init__()
        self.y_Ip = Var(ns, ps, bounds=(0, 1), within=Binary, initialize=0, doc='initial to final')


model = MyModel(ns=('inlet_liquid', 'inlet_vapor'), ps=('outlet_liquid', 'outlet_vapor'))
print([model.y_Ip, type(model.y_Ip)])  # IndexedVar
x = {k: 1 for k in model.y_Ip}  # ok
for (n, p) in model.y_Ip:  # "None" is not iterable
    print((n, p))
```

This reports `"None" is not iterable`, although both `Var` and `IndexedVar` implement `__iter__` through their parent `IndexedComponent` to return `self.keys()`.  Will this be fixed as well, or is that a separate issue?

**Comment by erictraut** [comment]
@alexchandel, that's because there is a code path that returns `None` from the iterator, and if you attempt to iterate over `None` to unpack `n` and `p`, it will crash. 

You are likely to get many false positive type errors when using a library that contains no type information, especially if the library is performing a lot of dynamic "magic" that cannot be determined by a static type analyzer. See my comment about about workarounds.

**Comment by debonte** [comment]
This issue has been fixed in prerelease version 2023.8.11, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023811-2-August-2023-PreRelease)
